text/0000-extern-impl.md

```diff
@@ -0,0 +1,624 @@
+even though it only cares about `Mysql`.
+```mermaid
+graph TB
+MysqlUser(MysqlUser) --> DBTrait --> Mysql & MongoDB & RocksDB
+```
+
+With this RFC, the implementation of `DBTrait` for each of the database crates
+would be distributed among multiple implementation crates. Downstream users
+would only need to take a dependency on the specific implemenetations they need.
+
+In this example, `MysqlUser` depends only on `DBTrait` and `DBMysql`, and is
+unaffected by the `DBMongo` and `DBRocks` implementations.
+```mermaid
+graph TB
+Bin(MysqlUser) --> DBMysql
+DBMysql & DBMongo & DBRocks -->|impl| DBTrait
+DBMysql --> Mysql
+DBMongo --> MongoDB
+DBRocks --> RocksDB
+```
+# Guide-level explanation (cargo)
+[guide-level-explanation]: #guide-level-explanation 
+
+Within a workspace, you can designate an "impl" relationship between two crates.
+For example, in your `my_impls/Cargo.toml` you can put:
+
+```
+[dependency]
+my_definitions = { version = "0.1", path = "../my_definitions", impl = True }
+```
+
+This means that your `my_impls` crate can add trait implementations using types
+or traits defined in `my_definitions`, almost as if they were defined in
+`my_definitions`. If you want to use those implementations in other crates,
+however, you must also add a dependency to both `my_impls` and `my_definitions`
+(or you could re-export the definitions in `my_impls`).
+
+You may only implement for types/traits actually defined in `my_definitions`
+itself. Re-exports don't count.
+
+The dependency *must* be a path dependency to another crate defined in the same
+workspace. Cargo will not allow you specify `impl = True` for other crates.
+
+Publishing packages with `impl = True` to crates.io is not allowed.
+
+# Guide-level explanation (rustc)
+
+In current Rust, when you specify in your build system that a crate depends on
+another, this is a "use" dependency - the dependee can use definitions in the
+dependant.
+
+This RFC defines a new dependency relationship: the "impl" dependency. The
+dependee can provide implementations for types and traits defined in the
+dependant.
+
+We extend the `rustc` `--extern` flag with an additional `impl` flag: `--extern
+impl:dbtrait=dbtrait.rlib`. This means that the crate currently being compiled
+with this option, is allowed to implement traits and methods for types defined
+in the `dbtrait` crate (the *defining* crate).
+
+For example:
+```rust
+//! Defining crate
+trait DBTrait {}
+```
+```rust
+//! Implementation crate, also depends on the `mysql` crate.
+use mysql::MySql;
+
+// Allowed because we're depending on `dbtrait` with `impl`
+impl dbtrait::DBTrait for MySql {}
+```
+
+In other words, the implementing crate is considered the same as the defining
+crate as far as the orphan rule goes. Note that you may only implement traits
+and types for crates directly defined in the defining crate &mdash; you cannot
+implement re-exported types.
```

Comment:
Not sure entirely what this was supposed to mean (maybe there's a typo in there?) but, I might suggest something like:

> Note that this only applies to items directly defined in defining crate, and not to any items defined in transitive deps, regardless of whether those transitive deps are impl deps or not.